### DISCUSSION TITLE
Folding atom pair count fingerprint

### DISCUSSION THREAD

**Discussion by Haonan-Zhang** [rdkit/rdkit#7223] [Q&A] [answered]
Hi,

I want to know how does rdkit fold a sparse atom pair count fingerprint into a fingerprint of given fpSize. I know that for the bit-vector fingerprint this is done by taking the modulo operation. But I find that this is not the case for the atom pair count fingerprint. To illustrate:
```
from rdkit.Chem import rdFingerprintGenerator
from rdkit import Chem

mol = Chem.MolFromSmiles('CCC')
fpgen = rdFingerprintGenerator.GetAtomPairGenerator(fpSize=2048)
fp = fpgen.GetCountFingerprint(mol)
print(fp.GetNonzeroElements())
```
{1311: 1, 1375: 2}
```
fp = fpgen.GetSparseCountFingerprint(mol)
print(fp.GetNonzeroElements())
```
{541730: 1, 558113: 2}
```
print(541730 % 2048)
print(558113 % 2048)
```
1058
1057
As you can see, the remainders are not the same. The problem is, I want to fold the BP fingerprint in the AtomPairs.Sheridan module (in the same way of AP fingerprint), which does not have the fpSize argument.

**Comment by greglandrum** [comment]
Hi @Haonan-Zhang,  with the atom pair and topological torsion fingerprints it is possible to generate a bit position for an atom environment (one atom pair or one torsion) directly from the environment itself. This is what you get when you call `GetSparseCountFingerprint()` or `GetSparseFingerprint()`. These are hashed a second time and then "folded" (by doing modulo) when you call `GetCountFingerprint()` or `GetFingerprint()`.

FYI, the Morgan fingerprints behave as you would expect: there is only one hashing step involved there:
```
In [7]: fpgen = rdFingerprintGenerator.GetMorganGenerator(fpSize=2048)

In [8]: fp = fpgen.GetCountFingerprint(mol)

In [9]: fp.GetNonzeroElements()
Out[9]: {80: 1, 294: 2, 1057: 2, 1344: 1}

In [10]: fp = fpgen.GetSparseCountFingerprint(mol)
    ...: [k%2048 for k in fp.GetNonzeroElements()]
Out[10]: [1344, 80, 1057, 294]
```

**Reply by Haonan-Zhang** [reply, marked as answer]
Hi Greg,
Thanks for your comment. Although I still don't get the hash function for the count fingerprint, I manage to create the hashed (folded) BP fingerprint by using the old fingerprint function:
```
from rdkit.Chem.rdMolDescriptors import GetHashedAtomPairFingerprint
from rdkit.Chem.AtomPairs import Sheridan

def typ_mol(mol):
    typMap = dict(CAT=1, ANI=2, POL=3, DON=4, ACC=5, HYD=6, OTH=7)
    typs = [typMap[x] for x in Sheridan.AssignPattyTypes(mol)]
    return typs

mol = Chem.MolFromSmiles('CCC')
ap = GetHashedAtomPairFingerprint(mol,2048, atomInvariants=typ_mol(mol))
print(ap.GetNonzeroElements())
```
I notice that GetAtomPairGenerator also has an atomInvariantsGenerator argument, but I don't know how to use it. So I will go with this walkaround for now.

**Comment by 1355wss** [comment]
I also want to know what chemical structure does each bit for atompair fingerprint represent when used GetCountFingerprint() to generate.

**Reply by greglandrum** [reply]
There's some info about how to do this in this blog post: https://greglandrum.github.io/rdkit-blog/posts/2023-01-18-fingerprint-generator-tutorial.html#additional-information-explaining-bits
